Loading

willrogers opened the pull request.

willrogers committed:
Color: use tuple for (r, g, b).

It only makes sense to specify these together.
Default to (0, 0, 0).

willrogers committed:
Merge branch 'colors' into loading

willrogers committed:
Add dynamic loading of color.def file.

willrogers committed:
Allow numbers in mangled variable names.

willrogers committed:
Better test name.

willrogers committed:
Load fonts from font.def file into opimodel.fonts namespace.

willrogers committed:
Correct import.

willrogers committed:
Handle both python 2 and 3 in test_colors.py.

willrogers committed:
Add mock to requirements.txt.

nickbattam commented on test/test_utils.py:
```diff
@@ -0,0 +1,11 @@
+                          ('Hello:', 'HELLO_'),
+                          ('Hello: Goodbye', 'HELLO_GOODBYE'),
+                          ('Dummy 1', 'DUMMY_1')])
+def test_mangle_name_handles_various_cases(input, output):
```
You need an extra test case for consecutive underscores as this is explicitly referenced in the mangle_name header

willrogers replied:
The colon-space sequence is converted into two underscores and reduced to one.

nickbattam replied:
You only know that because you wrote it. As `_` is identified a 'special' character (because non [A-Za-z0-9] characters gets converted into one) it would be worth having a test that demonstrates how it is handled.

nickbattam commented on opimodel/colors.py:
```diff
@@ -1,24 +1,34 @@
-    return def_colors
+                r, g, b = [int(x.strip(',')) for x in value.split()]
+                name = utils.mangle_name(key)
+                setattr(sys.modules[__name__], name, Color((r, g, b), key))
```
Should this log a warning if the name mangling reduces two colour names to the same mangled attribute name

willrogers replied:
... maybe ... but it is fairly unlikely.

nickbattam replied:
I'd log it - in the unlikely situation it happens it would be a very hard issue to trace.

nickbattam commented on opimodel/colors.py:
```diff
@@ -1,24 +1,34 @@
-GREEN = Color(0, 255, 0, 'Green')
-BLUE = Color(0, 0, 255, 'Blue')
+BLACK = Color((0, 0, 0), 'Black')
+RED = Color((255, 0, 0), 'Red')
```
What is the precedence of these colour constants over a similarly named colour in a parsed colour file? e.g. if I have defined `red=(192,0,0)`

willrogers replied:
The parsed file will overwrite the predefined ones.  Not sure what I think about this exactly.

nickbattam replied:
I think that's the right behaviour, unless you particularly need these primary colours to defined; if you do you could rename them `CSSGEN_RED` etc.
It's worth documenting that and/or putting a unittest in to demonstrate it

nickbattam commented on opimodel/utils.py:
```diff
@@ -0,0 +1,27 @@
+                continue
+            else:
+                char = '_'
+        last = char
```
this final statement is the same as the `if` block. Could the logic be rewritten:
```
if not char.isalpha() and not char.isdigit():
    if last == '_': 
        continue
    else:
        char = '_'
last = char
```

willrogers replied:
possibly let's check if the test still passes ...

nickbattam commented on opimodel/colors.py:
```diff
@@ -1,24 +1,34 @@
+    Parse the provided color.def file, create Color objects for each
+    defined color and attach them to the namespace of this module wth
+    names converted into appropriate constants by the
+    utils.mangle_name() function.
```
Somewhere the documentation needs to say parse should be called first.
Does someone calling the function really need to know it uses `utils.mangle_name`?

willrogers replied:
The user needs to know how the font names are specified in the file once the function is called.

nickbattam replied:
It would be more helpful to describe what the function is doing not just referencing the function. 
Saying 'appropriate constants' means I have to go and find/read the `utils.mangle_name` docs. Also, if you ever refactor the code that comment will be out-of-date.

willrogers replied:
On the other hand, if I ever change the behaviour of `utils.mangle_name()` then this comment would become incorrect.

nickbattam commented on opimodel/fonts.py:
```diff
@@ -28,9 +32,23 @@ def __eq__(self, other):
+    Parse the provided font.def file, create Font objects for each
+    defined font and attach them to the namespace of this module wth
+    names converted into appropriate constants by the
+    utils.mangle_name() function.
```
Same as for the colours parse comments: no need to mention `utils.mangle_name` and docs need to say when this should be called.

willrogers commented:
@nickbattam @mfurseman I am puzzled by the Travis test failures.  The same tests pass on Python 2.7 and Python 3.4 on my machine.

willrogers committed:
Minor logic simplification.

willrogers committed:
Remove predefined colors.

willrogers committed:
Fix test by adding explicit color.

willrogers committed:
Require mock version 2 or greater.

willrogers committed:
Refactor common behaviour from parsing font and color files.

Create utils.add_attr_to_module() function to attach variables to module
namespaces.

nickbattam commented on opimodel/colors.py:
```diff
@@ -24,5 +24,5 @@ def parse_css_color_file(filepath):
             if not line == '' and not line.startswith('#'):
```
`not line == ''` is the same as `if line` isn't it?
The same fragment is in the font file parser

nickbattam commented on opimodel/utils.py:
```diff
@@ -23,3 +23,18 @@ def mangle_name(name):
+    """
+    var = mangle_name(name)
+    if hasattr(module, var):
+        print('Warning: overwriting variable {} in module {}'.format(var, module))
```
Should this be a `log.warn` or `log.info` rather than a `print`?

willrogers committed:
Minor simplifications.
